### PR TITLE
Added Keccak256 to supported algorithms

### PR DESCRIPTION
**Pull request by scglenn**

### CODE REVIEW

File: src/algorithms/hash_algorithm.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+use crate::{prelude::*, Hasher};
+
+#[derive(Clone)]
+pub struct HashAlgorithm {}
+
+#[derive(Clone)]
+pub enum HashType {
+    Keccak256,
+    Sha256,
+}
```
Comment (KaiserKarel) [inline src/algorithms/hash_algorithm.rs]: Does Hashtype need to be a closed set?

File: src/hasher.rs [hunk 2]
```diff
@@ -43,7 +50,7 @@ pub trait Hasher: Clone {
 
     /// This associated function takes a slice of bytes and returns a hash of it.
     /// Used by `concat_and_hash` function to build a tree from concatenated hashes
-    fn hash(data: &[u8]) -> Self::Hash;
+    fn hash(data: &[u8], hash_type: HashType) -> Self::Hash;
```
Comment (KaiserKarel) [inline src/hasher.rs]: I feel this definition is incorrect. Keccak IS a Hasher; so what would happen if we did:

```
keccak::hash(b"hello", HashType::SHA3)
```

File: src/hasher.rs [hunk 3]
```diff
@@ -63,7 +70,7 @@ pub trait Hasher: Clone {
             Some(right_node) => {
                 let mut right_node_clone: Vec<u8> = (*right_node).into();
                 concatenated.append(&mut right_node_clone);
-                Self::hash(&concatenated)
+                Self::hash(&concatenated, HashType::Keccak256)
```
Comment (scglenn) [inline src/hasher.rs]: this a bug

Comment (scglenn) [reply src/hasher.rs]: fixed it

File: tests/common.rs [hunk 4]
```diff
@@ -70,36 +66,36 @@ impl MerkleProofTestCase {
...
 }
 
-pub fn setup_proof_test_cases() -> Vec<ProofTestCases> {
+pub fn setup_proof_test_cases<T: Hasher>() -> Vec<ProofTestCases<T>> {
     let max_case = [
         "a", "b", "c", "d", "e", "f", "g", "h", "k", "l", "m", "o", "p", "r", "s",
     ];
 
     max_case
-        .par_iter()
```
Comment (scglenn) [inline tests/common.rs:79]: idk what i need to do to allow for par_iter to work with the generics update

**Comment by seunlanlege** [comment]
Guys why not just pre-sort?

**Pull request merged**